spring-cloud-gcp-starters/spring-cloud-gcp-starter-logging/src/main/java/org/springframework/cloud/gcp/logging/autoconfig/StackdriverLoggingAutoConfiguration.java

```diff
@@ -20,24 +20,64 @@
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.logging.CompositeTraceIdExtractor;
 import org.springframework.cloud.gcp.logging.LoggingWebMvcConfigurer;
+import org.springframework.cloud.gcp.logging.TraceIdExtractor;
 import org.springframework.cloud.gcp.logging.TraceIdLoggingWebMvcInterceptor;
+import org.springframework.cloud.gcp.logging.XCloudTraceIdExtractor;
+import org.springframework.cloud.gcp.logging.ZipkinTraceIdExtractor;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Import;
 
 /**
  * @author Mike Eltsufin
+ * @author Chengyuan Zhao
  */
 @Configuration
 @ConditionalOnClass({ TraceIdLoggingWebMvcInterceptor.class, TraceLoggingEnhancer.class })
+@EnableConfigurationProperties({ StackdriverLoggingProperties.class })
 @ConditionalOnProperty(value = "spring.cloud.gcp.logging.enabled", matchIfMissing = true)
 @Import(LoggingWebMvcConfigurer.class)
 public class StackdriverLoggingAutoConfiguration {
 
 	@Bean
 	@ConditionalOnMissingBean
-	public TraceIdLoggingWebMvcInterceptor getLoggingWebMvcInterceptor() {
-		return new TraceIdLoggingWebMvcInterceptor();
+	public TraceIdLoggingWebMvcInterceptor loggingWebMvcInterceptor(
+			TraceIdExtractor extractor) {
+		return new TraceIdLoggingWebMvcInterceptor(extractor);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public TraceIdExtractor traceIdExtractor(
+			StackdriverLoggingProperties loggingProperties) {
+		TraceIdExtractorType type = loggingProperties.getExtractorType();
+		TraceIdExtractorType checkedCombination = type == null
+				? TraceIdExtractorType.XCLOUD_ZIPKIN
+				: type;
+		TraceIdExtractor extractor;
+		switch (checkedCombination) {
+		case XCLOUD:
+			extractor = new XCloudTraceIdExtractor();
+			break;
+		case ZIPKIN:
+			extractor = new ZipkinTraceIdExtractor();
+			break;
+		case ZIPKIN_XCLOUD:
+			extractor = new CompositeTraceIdExtractor(
+					new ZipkinTraceIdExtractor(), new XCloudTraceIdExtractor());
+			break;
+		case XCLOUD_ZIPKIN:
+			extractor = new CompositeTraceIdExtractor(
+					new XCloudTraceIdExtractor(), new ZipkinTraceIdExtractor());
+			break;
+		default:
+			extractor = new CompositeTraceIdExtractor(
```

Comment:
Since the default case is impossible when all cases are covered by the `switch`, I would just throw an `IllegalArgumentException` here to catch mistakes early.

Reply:
It won't compile without the default statement since Java isn't a dependently-typed language and the compiler doesn't know that all cases are covered  (and also cuz I don't assign `extractor` outside of the switch statement)

Reply:
I'm not suggesting to remove the default case, I'm just suggesting throwing an exception there.